### DISCUSSION TITLE
Is there a method to let external agents/applications call internal IDE agents?

### DISCUSSION THREAD

**Discussion by wescerliu** [eclipse-theia/theia#14575] [Q&A] [answered]
We have made a multiple-agent RAG system to enhance chat using thiea AI agent framework. Now we have other external applications or agents which would like to use this in IDE RAG Agent system as well. Let's say, send a question and recieve a final response, not in chat, but in a request format on the same server network.
We figured if we make a IDE instance into a headless service and make an api for external calls, probably will do, but not sure if it is actually feasible and seems too complicated.
So the question is, Is there or will there be a direct way of providing internal agents as a tool or a service to external calls.
Btw, there is a agent proxy block in the High Level Architecture of Theia AI diagram, but we are not sure what exactly is this and is it available now in the source code.
Thank you!

**Comment by sdirix** [comment, marked as answer]
>We have made a multiple-agent RAG system to enhance chat using thiea AI agent framework. Now we have other external applications or agents which would like to use this in IDE RAG Agent system as well. Let's say, send a question and recieve a final response, not in chat, but in a request format on the same server network.
We figured if we make a IDE instance into a headless service and make an api for external calls, probably will do, but not sure if it is actually feasible and seems too complicated.
So the question is, Is there or will there be a direct way of providing internal agents as a tool or a service to external calls.

Thank you for your question! Currently, there is no built-in functionality or planned feature in Theia to expose internal AI agents automatically via an API for external use.

However, the Theia backend is designed to be highly customizable. You can easily extend it to provide additional REST or WebSocket endpoints tailored to your specific requirements. By doing so, you can enable external tools or applications to interact with your multiple-agent RAG system in a structured way.

For example, you can start a Theia instance and modify its backend to include custom API endpoints that:

- Accept external requests (e.g., sending questions to your RAG agents).
- Forward the input to the desired agent(s).
- Return the final response to the external caller in the desired format.

See [here](https://github.com/eclipse-theia/theia/blob/52945586d4d7c037cc0c35b15a9e1804fccc8e66/packages/filesystem/src/node/download/file-download-endpoint.ts) for an example on how to configure an endpoint. In your case you could for example expose an `agent-request` endpoint.

> Btw, there is a agent proxy block in the High Level Architecture of Theia AI diagram, but we are not sure what exactly is this and is it available now in the source code.

Regarding the "Agent Proxy" block in the Theia AI framework's High-Level Architecture diagram: this is conceptual and intended to represent the mechanism by which agents can be orchestrated or communicated with. The idea is that an agent does not need to be fully implemented within Theia, instead a shallow "proxy agent" could be implemented instead which delegates to the actual agent running outside of Theia.

**Reply by wescerliu** [reply]
Great appreciation to your quick response and thorough answer, let's rock!
